File: klassen/projekt.py
# Definition der Klasse Projekt

import json


class Projekt:
    """Initialisierungs-Funktion der Klasse "Projekt"

    :param name: Name
    :param id: ProjektID
    :param beschr: Beschreibung
    :param ort: Ort
    :param lehrkraft: Lehrkraft
    :param maxAnzahl: Maximale Anzahl an Schülern
    :param minStufe: Niedrigste erlaubte Klassenstufe
    :param maxStufe: Höchste erlaubte Klassenstufe
    :param preis: Preis
    :returns: Projekt - Self

    """
    def __init__(self, name : str, id: int, beschr: str, ort: str, lehrkraft: str, maxAnzahl: int, minStufe: int,maxStufe: int, preis: int = 0):
        
        self.name = name
        self.id = id
        self.beschr = beschr
        self.ort = ort
        self.lehrkraft = lehrkraft
        self.maxAnzahl = maxAnzahl
        self.minStufe = minStufe
        self.maxStufe = maxStufe
        self.preis = preis
    # Objekt zu Dictionary konvertieren
    def toDict(self):
        newMap = {}
        newMap["name"] = self.name
        newMap["id"] = self.id
        newMap["beschr"] = self.beschr
        newMap["ort"] = self.ort
        newMap["lehrkraft"] = self.lehrkraft
        newMap["maxAnzahl"] = self.maxAnzahl
        newMap["minStufe"] = self.minStufe
        newMap["maxStufe"] = self.maxStufe
        newMap["preis"] = self.preis
        return newMap
        
    # Dictionary zu Objekt konvertieren
    def fromDict(inpDict: dict):

        return Projekt(        
            inpDict["name"],
            inpDict["id"],
            inpDict["beschr"], 
            inpDict["ort"],
            inpDict["lehrkraft"], 
            inpDict["maxAnzahl"],
            inpDict["minStufe"],
            inpDict["maxStufe"],
            inpDict["preis"]
            )
    # Liste an Dicts zu Liste an Objekten konvertieren
    def fromListOfDicts(inpList: list):
        neueListe = []
        for element in inpList:
            neueListe.append(Projekt.fromDict(element))
        return neueListe
    # Objekt zu String konvertieren
    def toString(self):
        return json.dumps(self.toDict())
    
    '''Diese Funktion erlaubt es dem 'print' Befehl die Klasse als Text zu schreiben
    
    '''
    def __str__(self):
        return json.dumps(self.toDict())

File: src/classes/projekt.ts
/**
 * Interface für die Typisierung der Rohdaten
 */
export interface ProjektData {
  name: string;
  id: number;
  beschr: string;
  ort: string;
  lehrkraft: string;
  maxAnzahl: number;
  minStufe: number;
  maxStufe: number;
  preis: number;
}

/**
 * Die Projekt-Klasse
 * Durch das "export" kann sie in .vue oder .ts Dateien importiert werden.
 */
export class Projekt implements ProjektData {
  constructor(
    public name: string,
    public id: number,
    public beschr: string,
    public ort: string,
    public lehrkraft: string,
    public maxAnzahl: number,
    public minStufe: number,
    public maxStufe: number,
    public preis: number = 0
  ) {}

  toDict(): ProjektData {
    return {
      name: this.name,
      id: this.id,
      beschr: this.beschr,
      ort: this.ort,
      lehrkraft: this.lehrkraft,
      maxAnzahl: this.maxAnzahl,
      minStufe: this.minStufe,
      maxStufe: this.maxStufe,
      preis: this.preis,
    };
  }

  static fromDict(inpDict: ProjektData): Projekt {
    return new Projekt(
      inpDict.name,
      inpDict.id,
      inpDict.beschr,
      inpDict.ort,
      inpDict.lehrkraft,
      inpDict.maxAnzahl,
      inpDict.minStufe,
      inpDict.maxStufe,
      inpDict.preis ?? 0
    );
  }
  static fromString(jsonString: string): Projekt {
    return Projekt.fromDict(JSON.parse(jsonString));
  }
  static fromListOfDicts(inpList: ProjektData[]): Projekt[] {
    return inpList.map((element) => Projekt.fromDict(element));
  }

  toString(): string {
    return JSON.stringify(this.toDict());
  }
}